### PR TITLE
[Offline Trader] Exploit Fix & Overflow Protection

### PR DESCRIPTION
**Pull request by Valorith**
## Summary

Fixes two bazaar purchase safety issues:

- **Exploit Fix:** rejects crafted bazaar-window purchase packets with invalid quantities or client-supplied prices before money, parcel, active-transaction, or world-messaging side effects occur.
- **Overflow Fix:** normalizes bazaar transaction caps to a safe 2,000,000,000 copper default and performs `price * quantity` checks in 64-bit arithmetic before funds are moved.

## Root Cause

The bazaar-window parcel purchase path trusted client-supplied `quantity` and `price`. A crafted request could use invalid quantity values or a mismatched/zero price, allowing the flow to pass buyer-side checks and reach seller/world handling with inconsistent transaction state.

Separately, the client-version transaction cap lookup could produce unsafe caps: unset client versions inherited zero, while RoF2 advertised a multi-trillion-copper limit that the money and packet paths do not safely support. One direct bazaar buy path also used arithmetic that could wrap before money moved.

## Changes

- Adds centralized bazaar purchase quantity validation.
- Rejects zero, negative-encoded, invalid stack, and invalid non-stack quantities.
- Validates client price against the server-side trader listing price and uses the listed price downstream.
- Adds a shared `BAZAAR_MAX_TRANSACTION_DEFAULT` of 2,000,000,000 copper.
- Normalizes unset lookup caps and RoF2 to the safe default.
- Uses explicit 64-bit multiplication before bazaar transaction cap checks.
- Adds unit coverage for bazaar purchase validators and transaction cap lookup values.

## Testing
Confirmed Fixed
image

## Validation

- `cmake --build build-codex --target tests --config Debug -- /m`
- `build-codex\bin\Debug\tests.exe` passed 132/132 tests.

---

> [!NOTE]
> **Medium Risk**
> Changes bazaar money limits and purchase validation on live trading paths; incorrect validation could block legitimate buys or leave edge cases in stackable quantity handling.
> 
> **Overview**
> **Bazaar transaction limits** are normalized so RoF2 and unset client-version lookups use a shared **2,000,000,000 copper** cap instead of a zero or multi-trillion value that the money/packet paths cannot safely honor. Direct and parcel bazaar buys now compute **`price * quantity` in 64-bit** before comparing against that cap.
> 
> **Purchase validation** is centralized in **`Bazaar::ValidatePurchaseQuantity`** and **`ValidatePurchasePrice`**, and the parcel/bazaar-window path rejects bad client-supplied quantity/price and **overwrites** successful requests with listed charges and cost from the trader record before funds move or world messaging runs. The in-person trader buy path tightens zero-price/zero-qty checks and uses the same cap lookup pattern.
> 
> **Tests** cover the new validators and that static lookups expose the safe default cap.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4fb72ce55c73b5bcbfe51ad4093f1b0fabf24d44. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: zone/trading.cpp [hunk 1]
```diff
@@ -2909,12 +2906,13 @@ void Client::BuyTraderItemFromBazaarWindow(const EQApplicationPacket *app)
 		in->item_unique_id
 	);
 
+	const uint64 max_transaction_value = EQ::constants::StaticLookup(ClientVersion())->BazaarMaxTransaction;
 	uint64 total_cost = static_cast<uint64>(in->price) * static_cast<uint64>(quantity);
-	if (total_cost > EQ::constants::StaticLookup(ClientVersion())->BazaarMaxTransaction) {
+	if (total_cost > max_transaction_value) {
 		Message(
 			Chat::Red,
```
[the remark this answers is not in the document]
Comment (Valorith) [reply zone/trading.cpp:2951]: Handled in the latest commits. `BuyTraderItemFromBazaarWindow` now rejects invalid purchase quantities before transaction side effects, and it also rejects zero or mismatched prices by validating the client price against the server-side trader listing price. The sanitized listed price is then used for the cap check and seller/world messaging.

Verified with `cmake --build build-codex --target tests --config Debug -- /m` and `build-codex\bin\Debug\tests.exe` (`132/132` passing).

---
<sub>Codex</sub>
